gateway/reverse_proxy.go

```diff
@@ -785,73 +790,126 @@ func (rt *TykRoundTripper) RoundTrip(r *http.Request) (*http.Response, error) {
 }
 
 const (
-	inMemNetworkName = "in-mem-network"
-	inMemNetworkType = "memu"
+	checkIdleMemConnInterval = 5 * time.Minute
+	maxIdleMemConnDuration   = -time.Minute
+	inMemNetworkName         = "in-mem-network"
+	inMemNetworkType         = "memu"
 )
 
-var (
-	memconnProvidersMtx sync.Mutex
-	// TODO: Find a proper way to clean that cache
-	memconnProvidersCache = make(map[string]*memconn.Provider)
-)
+type memConnProvider struct {
+	listener net.Listener
+	provider *memconn.Provider
+	lastUsed int64
+}
 
-func createMemconnProviderIfNeeded(handler http.Handler, r *http.Request) error {
-	memconnProvidersMtx.Lock()
-	defer memconnProvidersMtx.Unlock()
-	_, ok := memconnProvidersCache[r.Host]
+var memConnProviders = &struct {
+	mtx sync.Mutex
+	m   map[string]*memConnProvider
+}{
+	m: make(map[string]*memConnProvider),
+}
+
+// cleanIdleMemConnProvidersEagerly deletes unused memconn.Provider instances and closes
+// its listener.
+func cleanIdleMemConnProvidersEagerly(maxIdleDuration time.Duration) {
+	memConnProviders.mtx.Lock()
+	defer memConnProviders.mtx.Unlock()
+
+	for host, mp := range memConnProviders.m {
+		if time.Now().Add(maxIdleDuration).Unix() > mp.lastUsed {
```

Comment:
So, technically we're substracting maxIdleDuration here, which should be passed as an absolute/positive value. I'd prefer (-maxIdleDuration) here (semantics)